### ISSUE TITLE
Implement automatic model conversion and model serving image selection to support heterogeneous hardware platform

### ISSUE THREAD

**Issue by EnfangCui** [kubeedge/sedna#16]
**Function description:** Automatically select AI server (TFserving, PaddleServing, MindSpore Serving, Tensorflow lite, Paddle lite, MNN, OpenVINO, etc.) according to the hardware platform, and automatically perform model format conversion.

**Solution:**
Provided python SDK to realize the following functions:
1) Read the edge node hardware type label based on k8s python sdk and select model serving image. Edge node hardware type is labeled by kubectl as follows:
```
kubectl label node edge-node-1 hardware=Ascend
```
2) Deploy model conversion image for model conversion; 
3) Use Sedna k8s CRD for model serving image deployment.

**Comment by llhuii** [comment]
> Provided python SDK to realize the following functions:

Can we provide the other ways not only python SDK?

**Comment by EnfangCui** [comment]
> Can we provide the other ways not only python SDK?

@llhuii Yes, another alternative way is to provide a k8s operator. This may be a better way.

**Comment by ugvddm** [comment]
please refer to klever,  which have implemented  model conversion

**Comment by llhuii** [comment]
after simply digging to klever, its component klever-model-registry has built some [converter scripts](https://github.com/kleveross/klever-model-registry/tree/master/scripts/convert/converter).
@EnfangCui can you check it more deeper?

**Comment by EnfangCui** [comment]
> after simply digging to klever, its component klever-model-registry has built some [converter scripts](https://github.com/kleveross/klever-model-registry/tree/master/scripts/convert/converter).
> @EnfangCui can you check it deeper?

@llhuii OK, I will study it carefully.

**Comment by EnfangCui** [comment]
# Solution for model conversion
Add model job operator to provide functions related to the model operation, including:</br>
1. Model conversion: convert models to different formats to facilitate model deployment.
2. Model evaluation: evaluate the accuracy and accuracy loss of the model after conversion.
## Model Conversion
We use object storage schemes for model storage,  the open source solution MinIO and public cloud solutions are alternative. The object storage schemes used is based on the S3 protocol. When saving an model, you need to set the URL of the object storage server, the storage bucket, and the model file name. Specific examples are as follows:
1. MinIO
``` bash
apiVersion: sedna.io/v1alpha1
kind: ModelJob
metadata:
  name: modeljob-onnx-convert
  namespace: default
spec:
  storageURL: "http://121.37.140.33:9090/minio/" #URL of MinIO server
  bucket: "sedna-models" #storage bucket
  srcModel: "cnn_tensorflow" #The name of the source model
  dstModel: "cnn_onnx" #The name of the converted model
  conversion: #Specify the task as a model conversion task
      from: "saved_model" #Format of the source model
      to: "onnx" #Format of the converted model
```

2. Public Cloud
``` bash
apiVersion: sedna.io/v1alpha1
kind: ModelJob
metadata:
  name: modeljob-onnx-convert
  namespace: default
spec:
  storageURL: "obs.cn-north-4.myhuaweicloud.com"
  bucket: "sedna-models"
  srcModel: "cnn_tensorflow"
  dstModel: "cnn_onnx"
  conversion:
      from: "saved_model"
      to: "onnx"
```
**The detailed process of the model conversion job:**

![image](https://raw.githubusercontent.com/EnfangCui/sedna-proposal/main/imgs/model-convert.PNG)
## Model Evaluation
The converted model may have a loss of accuracy and needs to be evaluated. We provide CRD to evaluate the accuracy of the model.
``` bash
apiVersion: sedna.io/v1alpha1
kind: ModelJob
metadata:
  name: modeljob-eval
  namespace: default
spec:
  modelSpec:
    storageURL: "http://121.37.140.33:9090/minio/"
    bucket: "sedna-models"
    modelName: "cnn_tensorflow"
    modelFramework: "tensorflow"
  datasetSpec
    httpURL: "http://ai-atest.bj.bcebos.com/cifar-10-python.tar.gz"
  outputSpec: 
    bucket: "model-eval"
    fileName: "cnn_tensorflow_eval.csv"
```

**Comment by llhuii** [comment]
Hi, I have some questions:
1. From the Spec of ModelJob, only object storage is supported? it's not easy to extend other storage protocol since the top of spec has such `bucket` field.
2. Do srcModel/dstModel  just mean the name of model file? Or the sedna model crd?
3. Does storageURL mean the storage of srcModel or dstModel or both? If both, how can the user save the converted model to another position?
4. How to handle the ak/sk of minio?

**Comment by llhuii** [comment]
More questions:

> The converted model may have a loss of accuracy and needs to be evaluated

1. I'm curious that why having a loss, since it is just converting instead of pruning/quantization.
1. Does the user need to provide evalution code?
2. What metrics will the eval job output?
3. Missing `:` after `evaluation` and `datasetSpec` in `modeljob-eval` spec

**Comment by EnfangCui** [comment]
> Hi, I have some questions:
> 
> 1. From the Spec of ModelJob, only object storage is supported? it's not easy to extend other storage protocol since the top of spec has such `bucket` field.
> 2. Do srcModel/dstModel  just mean the name of model file? Or the sedna model crd?
> 3. Does storageURL mean the storage of srcModel or dstModel or both? If both, how can the user save the converted model to another position?
> 4. How to handle the ak/sk of minio?

@llhuii 
Answer 1:  I modified the design of the CRD and added the "storageMethod" field for users to select the storage method.
Answer 2: I gave a more detailed explanation to model name . You can check my updated solution.
Answer 3: I modified the design of the CRD and separate the storage methods of source model and destination model.
Answer 4: Deploy the operator in the way of deployment, and write AK / SK to the env of the operator container. When the operator creates pod for model convertion, AK / SK  in operator env will be read and written to env of the container for model pulling.

**Comment by llhuii** [comment]
Thanks, clearer.
More questions:
>       storageMethod: "s3"
>      storageURL: "http://121.37.140.33:9090/minio/" #URL of storage server
>      modelName: "cnn_onnx.tar.gz" #The file name of the converted model

How about just one storage URL `http://121.37.140.33:9090/minio/cnn_onnx.tar.gz`? note: this sample has inconsistent protocol (one is s3, one is http)

> Answer 4: Deploy the operator in the way of deployment, and write AK / SK to the env of the operator container. When the operator creates pod for model convertion, AK / SK in operator env will be read and written to env of the container for model pulling.

This way restricts only one minio storage support in the cluster. If the user want to use other user's shared model in other s3, this way injects the wrong ak/sk leading to error downloading.

**Comment by EnfangCui** [comment]
> How about just one storage URL `http://121.37.140.33:9090/minio/cnn_onnx.tar.gz`? note: this sample has inconsistent protocol (one is s3, one is http)

Answer: Yes, just one storage URL is okay, which is more convenient.

> This way restricts only one minio storage support in the cluster. If the user want to use other user's shared model in other s3, this way injects the wrong ak/sk leading to error downloading.

Answer: How  about use k8s Secret to manage secret keys? I modified the design of the CRD as follows: 

``` bash
apiVersion: sedna.io/v1alpha1
kind: ModelJob
metadata:
  name: modeljob-onnx-convert
  namespace: default
spec:
  conversion:
    srcModelSpec:
      storageMethod: "s3" # s3, http, https or hdfs
      storageURL: "http://121.37.140.33:9090/minio/cnn_tensorflow.tar.gz" #URL of the source model
      accessKeyRef:
          name: minio_secret
          key: accessKey
      secretKeyRef:
          name: minio_secret
          key: secretKey
    dstModelSpec:
      storageMethod: "s3"
      storageURL: "http://121.37.140.33:9090/minio/cnn_onnx.tar.gz" #URL of the converted model
      accessKeyRef:
          name: minio_secret
          key: accessKey
      secretKeyRef:
          name: minio_secret
          key: secretKey
    formatSpec:
      from: "saved_model" #Format of the source model
      to: "onnx" #Format of the converted model
```
The secret key is managed by k8s Secret:

``` bash
apiVersion: v1
kind: Secret
metadata:
  name: minio_secret
  namespace: default
type: Opaque
data:
  accessKey: "ZW5mYW5nY3Vp"
  secretKey: "ZW5mYW5nY3VpMTIzNDU="
```

**Comment by EnfangCui** [comment]
# Overall Solution
## Design Details

### CRD API Group and Version
The `HeteroInferenceService` CRD will be namespace-scoped.
The tables below summarize the group, kind and API version details for the CRD.

* HeteroInferenceService

| Field                 | Description             |
|-----------------------|-------------------------|
|Group                  | sedna.io     |
|APIVersion             | v1alpha1                |
|Kind                   | HeteroInferenceService   |

### Flow of Heterogeneous Inference
- The flow of heterogeneous inference service creation:
![image](https://raw.githubusercontent.com/EnfangCui/sedna-proposal/main/imgs/hetero-inference-flow-creation.png)
